apps/dashboard/src/utils/liquid-autocomplete.ts

```diff
@@ -149,73 +146,56 @@ function getFilterCompletions(afterPipe: string): CompletionOption[] {
-  return VALID_DYNAMIC_PATHS.some((path) =>
-    typeof path === 'string' ? searchText.startsWith(path) : path.test(searchText)
-  );
-}
-
-function validateSubscriberField(searchText: string, matches: LiquidVariable[]): LiquidVariable[] {
-  const parts = searchText.split('.');
-
-  if (parts.length === 2 && parts[0] === 'subscriber') {
-    if (!matches.some((v) => v.name === searchText)) {
-      return [];
-    }
-  }
-
-  return matches;
-}
-
-function validateStepId(searchText: string, variables: LiquidVariable[]): boolean {
-  if (!searchText.startsWith('steps.')) return true;
-
-  const stepMatch = searchText.match(/^steps\.([^.]+)/);
-  if (!stepMatch) return true;
-
-  const stepId = stepMatch[1];
-  return variables.some((v) => v.name.startsWith(`steps.${stepId}.`));
-}
-
 function getMatchingVariables(searchText: string, variables: LiquidVariable[]): LiquidVariable[] {
   if (!searchText) return variables;
 
   const searchLower = searchText.toLowerCase();
 
-  // Handle root prefixes and their partials
-  for (const [root, prefix] of Object.entries(ROOT_PREFIXES)) {
-    if (searchLower.startsWith(root) || root.startsWith(searchLower)) {
-      let matches = variables.filter((v) => v.name.startsWith(prefix));
-
-      // Special handling for subscriber fields
-      if (prefix === 'subscriber.') {
-        matches = validateSubscriberField(searchText, matches);
-      }
-
-      // Allow new paths for dynamic paths
-      if (isValidDynamicPath(searchText)) {
-        if (!matches.some((v) => v.name === searchText)) {
-          matches.push({ name: searchText } as LiquidVariable);
-        }
-      }
-
-      return matches;
-    }
-  }
-
   // Handle dot endings
   if (searchText.endsWith('.')) {
     const prefix = searchText.slice(0, -1);
     return variables.filter((v) => v.name.startsWith(prefix));
   }
 
-  // Validate step ID exists
-  if (!validateStepId(searchText, variables)) {
-    return [];
-  }
+  // Filter jit step namespaces out of the returned variables from the server
+  const stepPayloadNamespaces = variables.reduce<string[]>((acc, variableItem) => {
+    const match = variableItem.name.match(STEP_PAYLOAD_REGEX);
+
+    if (match) {
+      acc.push(match[0]);
+    }
+
+    return acc;
+  }, []);
+
+  // Create JIT variables based on the search text e.g. payload.foo, subscriber.data.foo, steps.digest-step.events[0].payload.foo
+  const jitVariables = [PAYLOAD_NAMESPACE, SUBSCRIBER_DATA_NAMESPACE, ...stepPayloadNamespaces].reduce<
+    LiquidVariable[]
+  >((acc, namespace) => {
+    if (searchText.startsWith(namespace) && searchText !== namespace) {
+      // Ensure that if the user types payload.foo the first suggestion is payload.foo
+      acc.push({ name: searchText, type: 'variable' });
+    } else if (!searchText.startsWith(namespace)) {
+      // For all other values, suggest payload.whatever, subscriber.data.whatever
+      acc.push({
+        name: `${namespace}.${searchLower.trim()}`,
```

Comment:
we shouldn't use `searchLower` as in the result it will suggest me wrong variable name; for example I have variable in my payload `firstName`, so I typed `{{firstName}}`, but in the dropdown it suggests `payload.firstname`

![Screenshot 2025-05-08 at 10 05 16](https://github.com/user-attachments/assets/59669a0e-3386-459d-996e-38c11cc29983)

Reply:
I love the attention to detail. Good catch.